### PR TITLE
Improve NEON32 encoding speed

### PR DESCRIPTION
**Pull request by mayeut**
The enc_translate function now uses the same tricks as the SSSE3 one.

Speed-up on `iPhone SE` using `Apple LLVM version 8.0.0 (clang-800.0.38)`
NEON32: +33% compared to previous version

Full results before & after modifications follows.
Before:
```
Filling buffer with 10.0 MB of random data...
Testing with buffer size 10 MB, fastest of 100 * 1
NEON32	encode	2174.78 MB/sec
NEON32	decode	1408.59 MB/sec
plain	encode	1004.75 MB/sec
plain	decode	706.02 MB/sec
Testing with buffer size 1 MB, fastest of 100 * 10
NEON32	encode	2191.70 MB/sec
NEON32	decode	1413.05 MB/sec
plain	encode	1043.34 MB/sec
plain	decode	708.30 MB/sec
Testing with buffer size 100 KB, fastest of 100 * 100
NEON32	encode	2189.48 MB/sec
NEON32	decode	1411.65 MB/sec
plain	encode	1028.19 MB/sec
plain	decode	713.58 MB/sec
Testing with buffer size 10 KB, fastest of 1000 * 100
NEON32	encode	2205.67 MB/sec
NEON32	decode	1418.22 MB/sec
plain	encode	1053.66 MB/sec
plain	decode	719.94 MB/sec
Testing with buffer size 1 KB, fastest of 1000 * 1000
NEON32	encode	2057.55 MB/sec
NEON32	decode	1312.66 MB/sec
plain	encode	1021.82 MB/sec
plain	decode	705.21 MB/sec
```
After:
```
Filling buffer with 10.0 MB of random data...
Testing with buffer size 10 MB, fastest of 100 * 1
NEON32	encode	2940.38 MB/sec
NEON32	decode	1427.19 MB/sec
plain	encode	1055.70 MB/sec
plain	decode	721.25 MB/sec
Testing with buffer size 1 MB, fastest of 100 * 10
NEON32	encode	2943.92 MB/sec
NEON32	decode	1428.84 MB/sec
plain	encode	1054.56 MB/sec
plain	decode	722.47 MB/sec
Testing with buffer size 100 KB, fastest of 100 * 100
NEON32	encode	2937.40 MB/sec
NEON32	decode	1425.56 MB/sec
plain	encode	1052.89 MB/sec
plain	decode	722.08 MB/sec
Testing with buffer size 10 KB, fastest of 1000 * 100
NEON32	encode	2927.49 MB/sec
NEON32	decode	1418.48 MB/sec
plain	encode	1053.09 MB/sec
plain	decode	725.24 MB/sec
Testing with buffer size 1 KB, fastest of 1000 * 1000
NEON32	encode	2685.02 MB/sec
NEON32	decode	1334.63 MB/sec
plain	encode	1016.33 MB/sec
plain	decode	710.38 MB/sec
```

### CODE REVIEW

**Comment by mayeut** [comment]
A check on another device and/or with another compiler might be a good idea before merging this in.

**Comment by aklomp** [comment]
Tested on the Raspberry 2 with `Debian clang version 3.0-6.2` shows the same result:

```
Filling buffer with 10.0 MB of random data...
Testing with buffer size 10 MB, fastest of 10 * 1
NEON32  encode  156.67 MB/sec
NEON32  decode  117.03 MB/sec
plain   encode  103.30 MB/sec
plain   decode  87.69 MB/sec
Testing with buffer size 1 MB, fastest of 10 * 10
NEON32  encode  156.82 MB/sec
NEON32  decode  117.38 MB/sec
plain   encode  103.32 MB/sec
plain   decode  87.83 MB/sec
Testing with buffer size 100 KB, fastest of 10 * 100
NEON32  encode  180.51 MB/sec
NEON32  decode  128.36 MB/sec
plain   encode  109.30 MB/sec
plain   decode  93.73 MB/sec
Testing with buffer size 10 KB, fastest of 100 * 100
NEON32  encode  182.34 MB/sec
NEON32  decode  129.23 MB/sec
plain   encode  110.00 MB/sec
plain   decode  94.15 MB/sec
Testing with buffer size 1 KB, fastest of 100 * 1000
NEON32  encode  176.32 MB/sec
NEON32  decode  126.49 MB/sec
```

```
Filling buffer with 10.0 MB of random data...
Testing with buffer size 10 MB, fastest of 10 * 1
NEON32  encode  188.33 MB/sec
NEON32  decode  116.03 MB/sec
plain   encode  102.33 MB/sec
plain   decode  86.96 MB/sec
Testing with buffer size 1 MB, fastest of 10 * 10
NEON32  encode  188.29 MB/sec
NEON32  decode  116.21 MB/sec
plain   encode  102.26 MB/sec
plain   decode  87.06 MB/sec
Testing with buffer size 100 KB, fastest of 10 * 100
NEON32  encode  244.81 MB/sec
NEON32  decode  128.04 MB/sec
plain   encode  108.98 MB/sec
plain   decode  93.48 MB/sec
Testing with buffer size 10 KB, fastest of 100 * 100
NEON32  encode  247.87 MB/sec
NEON32  decode  129.04 MB/sec
plain   encode  109.76 MB/sec
plain   decode  93.93 MB/sec
Testing with buffer size 1 KB, fastest of 100 * 1000
NEON32  encode  236.35 MB/sec
NEON32  decode  126.50 MB/sec
plain   encode  108.45 MB/sec
plain   decode  92.85 MB/sec
```

**Comment by aklomp** [comment]
I tried to compile with Raspbian's stock GCC-4.6.2, but it miscompiles the codec due to a known  bug (long since fixed, but not in Raspbian-land).

**Comment by aklomp** [comment]
It would be interesting to benchmark the new SIMD encoding stuff on NEON64 against the [native LUT-based codec](https://github.com/aklomp/base64/blob/master/lib/enc/neon64.c) for that arch. The codec (which I'm quite proud of) should be very fast since it uses very few instructions and processes 48 bytes per iteration. However, I was never able to benchmark it because I don't have access to NEON64 hardware (just an emulator: Qemu).

File: lib/codec_neon32.c [hunk 1]
```diff
@@ -39,68 +39,70 @@ enc_reshuffle (uint8x16x3_t in)
 	return out;
 }
 
+static inline uint8x16_t
```
Comment (aklomp) [inline lib/codec_neon32.c]: Trailing whitespace...

File: lib/codec_neon32.c [hunk 2]
```diff
@@ -39,68 +39,70 @@ enc_reshuffle (uint8x16x3_t in)
...
+	indices.val[2] = vsubq_u8(indices.val[2], mask.val[2]);
+	indices.val[3] = vsubq_u8(indices.val[3], mask.val[3]);
+
+	// lookup delta values
+	delta.val[0] = vqtbl1q_u8(lut, indices.val[0]);
+	delta.val[1] = vqtbl1q_u8(lut, indices.val[1]);
+	delta.val[2] = vqtbl1q_u8(lut, indices.val[2]);
+	delta.val[3] = vqtbl1q_u8(lut, indices.val[3]);
+
+
```
Comment (aklomp) [inline lib/codec_neon32.c]: Superfluous newline

**Comment by mayeut** [comment]
I did that on the same `iPhone SE`.
Don't look at the decode part, it's been modified (is now faster but I still have some ideas).
The existing encoding for NEON64 is faster than the implementation proposed for NEON32 here (I don't have the figures at hand, but remember what already exists for NEON64 is faster than this on arm64). Well, at least on that specific hardware but given the two implementations uses LUT, that should remain true whatever the hardware (LUT instruction can be quite slow on  some hardware).

```
Filling buffer with 10.0 MB of random data...
Testing with buffer size 10 MB, fastest of 100 * 1
NEON64	encode	4097.03 MB/sec
NEON64	decode	2310.51 MB/sec
plain	encode	1061.21 MB/sec
plain	decode	1235.06 MB/sec
Testing with buffer size 1 MB, fastest of 100 * 10
NEON64	encode	6553.44 MB/sec
NEON64	decode	2311.29 MB/sec
plain	encode	1062.54 MB/sec
plain	decode	1233.68 MB/sec
Testing with buffer size 100 KB, fastest of 100 * 100
NEON64	encode	6533.10 MB/sec
NEON64	decode	2309.77 MB/sec
plain	encode	1063.06 MB/sec
plain	decode	1232.53 MB/sec
Testing with buffer size 10 KB, fastest of 1000 * 100
NEON64	encode	6499.58 MB/sec
NEON64	decode	2307.98 MB/sec
plain	encode	1087.59 MB/sec
plain	decode	1237.40 MB/sec
Testing with buffer size 1 KB, fastest of 1000 * 1000
NEON64	encode	5840.41 MB/sec
NEON64	decode	2273.28 MB/sec
plain	encode	1059.23 MB/sec
plain	decode	1215.20 MB/sec
```

**Comment by mayeut** [comment]
I pushed the style fixes.

**Pull request merged**